src/rpc/stratum/dispatch.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright 2018 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::error::Error as StdError;
+use std::fmt;
+use std::marker::PhantomData;
+
+use futures::{future, Future, IntoFuture};
+use serde_json::Value as JsonValue;
+
+use super::error::{Error, Never};
+
+pub type Result = Box<Future<Item = Option<JsonValue>, Error = Error> + Send>;
+
+pub trait Dispatch {
+    type Error: Into<Box<StdError + Send + Sync>>;
+    type Future: Future<Item = Option<JsonValue>, Error = Self::Error>;
+
+    fn call(&mut self, req: JsonValue) -> Self::Future;
+}
+
+pub struct DispatchFn<F> {
+    f: F,
+    _req: PhantomData<fn(JsonValue)>,
```

Comment:
Why is it needed?

Reply:
I will look a little more about the PhantomData.